src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp

```diff
@@ -190,43 +190,94 @@ namespace
           : m_renderer(renderer)
           , m_active_camera(camera)
         {
+            m_callback_key = GetISceneEventManager()->RegisterCallback(this, false, 100, true);
         }
 
-        virtual void ControllerOtherEvent(NodeKeyTab& nodes) override
+        virtual ~SceneChangeCallback()
+        {
+            GetISceneEventManager()->UnRegisterCallback(m_callback_key);
+        }
+
+        virtual void ModelOtherEvent(NodeKeyTab& nodes) override
         {
             if (m_active_camera == nullptr || m_renderer == nullptr)
                 return;
-            
+
             for (int i = 0, e = nodes.Count(); i < e; ++i)
             {
                 if (NodeEventNamespace::GetNodeByKey(nodes[i]) == m_active_camera)
                 {
-                    m_renderer->update_camera_transform(m_active_camera);
+                    m_renderer->update_camera_object(m_active_camera);
                     m_renderer->get_render_session()->reininitialize_render();
                     break;
                 }
             }
         }
 
-        virtual void ModelOtherEvent(NodeKeyTab& nodes) override
+      private:
+        SceneEventNamespace::CallbackKey    m_callback_key;
+        AppleseedInteractiveRender*         m_renderer;
+        INode*                              m_active_camera;
+    };
+
+    class ViewportCallback 
+      : public RedrawViewsCallback
+    {
+      public:
+        ViewportCallback()
+          : m_current_view(nullptr)
+          , m_last_fov(0)
+          , m_last_timer(0)
+          , m_max_hwnd(GetCOREInterface()->GetMAXHWnd())
         {
-            if (m_active_camera == nullptr || m_renderer == nullptr)
-                return;
+            m_last_mat.IdentityMatrix();
 
-            for (int i = 0, e = nodes.Count(); i < e; ++i)
+            GetCOREInterface()->RegisterRedrawViewsCallback(this);
+        }
+
+        virtual ~ViewportCallback()
+        {
+            GetCOREInterface()->UnRegisterRedrawViewsCallback(this);
+        }
+
+        virtual void proc(Interface *ip) override
+        {
+            ViewExp& view_exp = ip->GetActiveViewExp();
+            if (m_current_view == nullptr)
+                m_current_view = &view_exp;
+
+            if (view_exp.IsAlive())
             {
-                if (NodeEventNamespace::GetNodeByKey(nodes[i]) == m_active_camera)
+                ViewExp13* vp13 = reinterpret_cast<ViewExp13*>(view_exp.Execute(ViewExp::kEXECUTE_GET_VIEWEXP_13));
+                    
+                Matrix3 curr_mat;
+                vp13->GetAffineTM(curr_mat);
+                if (m_last_mat.IsIdentity())
```

Comment:
I find this strange. You're already updating `m_last_mat` and `m_last_fov` when you detect a change and create a timer below.

Reply:
I thought they should be initialized first to avoid first changed just because _last_ vars are uninitialized. But it's probably not big deal, I'll remove this part.